Modules/itertoolsmodule.c

```diff
@@ -3454,12 +3454,81 @@ count_repr(countobject *lz)
                                 lz->long_cnt, lz->long_step);
 }
 
+static PyObject *
+count_peek(countobject *lz, PyObject *Py_UNUSED(ignored))
+{
+    if (lz->cnt != PY_SSIZE_T_MAX) {
+        return PyLong_FromSsize_t(lz->cnt);
+    } else {
+        return lz->long_cnt;
+    }
+}
+
+static PyObject *
+count_consume(countobject *lz, PyObject *seq)
+{
+    PyObject *it, *item;
+    PyObject *(*iternext)(PyObject *);
+
+    it = PyObject_GetIter(seq);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    iternext = *Py_TYPE(it)->tp_iternext;
+    while ((item = iternext(it)) != NULL) {
```

Comment:
I'm curious about the need for this pointer copy. Couldn't this be refactored to:
```suggestion
    while ((item = Py_TYPE(it)->tp_iternext(it)) != NULL) {
```

Although, you aren't really supposed to use `tp_iternext` directly, AFAIK. The proper call here would be to `PyIter_Next`.

Reply:
This is the most performant option, which can be used in certain cases. There is a fair amount of this pattern in existing code. See `collections` and `itertools`.

I think there should be a more formal official pattern to do this, but this is another topic. See: https://github.com/python/cpython/pull/121408 for brief discussions about it.

Even though there doesn't seem to be a clear consensus about it this case is a very simple one and is close to https://github.com/python/cpython/blob/9fc1c992d6fcea0b7558c581846eef6bdd811f6c/Modules/_collectionsmodule.c#L448, where there aren't any good reasons not to have it faster.

Reply:
Actually, this is mainly used by modules maintained by Raymond, which are `collections` and `itertools` (and those are also the most performance-critical).